### ISSUE TITLE
Credit memo for orders with zero TotalPaid

### ISSUE THREAD

**Issue by klakovsky** [magento/magento2#13153]
### Preconditions

1. Magento 2.3-develop
2. Create order with zero Amount (all Grand Total, Total Paid, Total Refunded, Total Due, Subtotal, Shipping & Handling, Tax are zero ) in status = "Pending" and 1 SKU qty = 1.

### Steps to reproduce
1. First Set tax calculation setting according to this screenshot
![57360176-14df4180-71ac-11e9-874b-6dcdd2a1f1d1](https://user-images.githubusercontent.com/50043226/57431130-642a7e00-724f-11e9-8cba-cfe0f1e314c0.png)
2. then Create a Cart price rule.
![Screenshot (5)](https://user-images.githubusercontent.com/50043226/57431290-f03ca580-724f-11e9-9c99-b7015d0d1eeb.png)
=> Total discount: $104.00
3. Tax percentage: 7% (all product prices on our store include tax, and our flat rate shipping fee includes tax as well)
4. Create a bundle product according to this screenshot
![2019-05-09-11-36-i-13153-2-3-develop instances magento-community engineering](https://user-images.githubusercontent.com/50043226/57431453-7f49bd80-7250-11e9-81fd-814b40720bbf.png)
5. Go to frontend, Add bundle product into the cart 
![Screenshot (6)](https://user-images.githubusercontent.com/50043226/57431702-3e05dd80-7251-11e9-87d6-6967d3f24375.png)
6. Go to checkout
7. Continue as guest or as authenticated user and place order
8. Create Invoice and now Order is status = "Processing"
9. Now create Credit memo and refund and return product
### Expected result
1. Credit memo's grand total should be zero.
![Screenshot (8)](https://user-images.githubusercontent.com/50043226/57432276-05ff9a00-7253-11e9-869e-6b66db1531e1.png)

### Actual result
1. I'm seeing a negative dollar value in the credit memo
![Screenshot (7)](https://user-images.githubusercontent.com/50043226/57432092-7bb73600-7252-11e9-8538-4e00524397c8.png)

-----------------------------------------------------------------------------------------------------------------------------------------

**Comment by magento-engcom-team** [comment]
@klakovskiy, thank you for your report.
We've acknowledged the issue and added to our backlog.

**Comment by viral-wagento** [comment]
@klakovskiy 

I think, Once the invoice/shipment is generated, you can not cancel order, only option available is to refund that order, if customer do not want to proceed for ordered items.

And when we place an order of Zero amount, there is no need to refund (or create credit memo) that order.
Please correct if any.

**Comment by AndrewPenry** [comment]
@viral-wagento 

Creating a credit memo is the way you return the items to stock.

**Comment by viral-wagento** [comment]
@AndrewPenry 
Yes, that can be the case, if we are managing stock to magento system.
I am working on this.

**Comment by viral-wagento** [comment]
@magento-engcom-team @ishakhsuvarov 
I am working on this. Please assign me.

**Comment by ishakhsuvarov** [comment]
@viral-wagento Please accept organization invite here https://github.com/orgs/magento/invitation 
You will be able to self-assign it afterwards.
Thank you

**Comment by franckgarnier21** [comment]
@AndrewPenry Did you find a workaround to manage the items return to stock with an order with zero amount ?

**Comment by AndrewPenry** [comment]
@franckgarnier21 No. It happens very rarely to us (maybe once per month at most), so we've just been dealing with it manually. We sync our store with our ERP, so we have just been making the credit memo in our ERP manually, and pushing a stock update back out to Magento. If we weren't using an ERP, I'd guess we'd just manually correct the stock in Magento when this occurs.

**Comment by Gaelle-MLR** [comment]
@AndrewPenry How do you cancel the order? Even if we manage the stock in the ERP, the sale is still "Processing" and we need it to to be "Closed"

**Comment by AndrewPenry** [comment]
@Gaelle-MLR Sorry, but we haven't found a way to do that without just editing the DB. If this were occurring more often, I'd probably try to program a module to fix it. But because the issue is so rare for us, I can't justify the hours. 

I guess one option would be to manually adjust your stock and then completely delete the order using a module like ibnab/magento2-delete-orders. Of course, that means you no longer have a record of what transpired, but it's a thought.

**Comment by wojtekn** [comment]
I also spot this issue. Fix backported from 2.3 to 2.2 seems to fix it, however, only in CE code. Use case in EE is still broken due to \Magento\CustomerBalance\Plugin\CreditMemoResolver::afterCanCreditmemo method behavior.

Also, it looks that after applying the fix "Credit Memo" button appears on order page but it's still not visible on the invoice page. It's caused by different conditions checked in both blocks: order block checks only for $order->canCreditmemo() but invoice block adds a few more checks.

**Comment by simonmaass** [comment]
is this fixed in 2.3? are you able to issue a credit memo for a zero invoice?

**Comment by mystix** [comment]
~Still not fixed in M2.3.1 😞~

I stand corrected. In M2.3.1, this configuration needs to be enabled:
![Screenshot 2019-05-07 at 2 56 00 PM](https://user-images.githubusercontent.com/174070/57278707-d0837100-70d9-11e9-9f01-6433cceea0c9.png)

(found under Admin > Stores > Configuration > Sales > Sales)

---

That said, when refunding a zero subtotal order, I'm seeing a negative dollar value in the credit memo, and the order status changes to "Complete" instead of "Closed". 🤦‍♂

**Comment by simonmaass** [comment]
you are right with the order status being "complete" - i can confirm... The credit memo on the other hand is just 0€ and looks fine

**Comment by mystix** [comment]
> you are right with the order status being "complete" - i can confirm... The credit memo on the other hand is just 0€ and looks fine

are you on M2.3.1 as well?

**Comment by simonmaass** [comment]
yes

**Comment by ghost** [comment]
Issue duplicated https://github.com/magento/magento2/issues/22762

**Comment by mystix** [comment]
> The credit memo on the other hand is just 0€ and looks fine

Are your tax settings like so?
![Screenshot 2019-05-08 at 4 11 52 PM](https://user-images.githubusercontent.com/174070/57360176-14df4180-71ac-11e9-874b-6dcdd2a1f1d1.png)

Seems like refund calculations for zero subtotal orders are messed up when
- the `Apply Discount On Prices` config is set to `Including Tax`
- order has not been shipped
- a credit memo is created for such an order

Instead of a $0 grand total for the credit memo, I'm seeing a total refund value of -(tax %) e.g. 
- total order value before discount: $104.00
- total discount: $104.00
- tax percentage: 7% (all product prices on our store include tax, and our flat rate shipping fee includes tax as well)
- credit memo grand total: -(104 / 107 * 7) = -$6.80

**Comment by simonmaass** [comment]
here are our tax settings:
![tax](https://user-images.githubusercontent.com/32599710/57361059-e5b0dc00-717b-11e9-8130-95890aaa98e5.png)

but the way u describe it - the order has a discount and there for the final price is 0€?

**Comment by mystix** [comment]
This is what we're seeing when we create a credit memo for a zero subtotal order and those settings from my previous screenshot:
(NOTE: our tax rate is 7%)

![Screenshot 2019-05-08 at 5 41 09 PM](https://user-images.githubusercontent.com/174070/57365877-802f1080-71b8-11e9-97ff-1f9c2e63716c.png)

Traced it down to the `collect()` method in `vendor/magento/module-sales/Model/Order/Creditmemo/Total/Discount.php`, where grand total (i.e. without tax) < discount amount (i.e. with tax) for whatever strange reason I have yet to figure out:
```
        $creditmemo->setGrandTotal($creditmemo->getGrandTotal() - $totalDiscountAmount);
        $creditmemo->setBaseGrandTotal($creditmemo->getBaseGrandTotal() - $baseTotalDiscountAmount);
```
which leads to the negative grand total value in the screenshot.

**Comment by simonmaass** [comment]
i tried to reproduce but works...

Created order with discount to the total order amount -> so zero total
![order](https://user-images.githubusercontent.com/32599710/57366969-65dd3e80-7188-11e9-9539-da42d9e79d98.png)

then created new zero credit memo
![new_memo](https://user-images.githubusercontent.com/32599710/57367004-7a213b80-7188-11e9-8339-e819601c609f.png)

**Comment by mystix** [comment]
Found the problem:

I tested with a **_simple product_** as you did, and the credit memo's grand total is correct as you've said.

Try creating a credit memo for a zero subtotal order containing a **_bundle product_** though, and things go out of whack:

![Screenshot 2019-05-08 at 11 09 42 PM](https://user-images.githubusercontent.com/174070/57386166-66a4bd80-71e6-11e9-84ac-017f9a57140b.png)

**Comment by ghost** [comment]
@webkul-deepakkumar please test this also check whether it is same as https://github.com/magento/magento2/issues/22762

**Comment by deepakk-webkul** [comment]
@magento-engcom-team give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @webkul-deepakkumar. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @webkul-deepakkumar, here is your [Magento instance](https://i-13153-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-13153-2-3-develop.instances.magento-community.engineering/admin](https://i-13153-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by deepakk-webkul** [comment]
I was able to reproduce this issue on vanilla Magento instance

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @webkul-deepakkumar
Thank you for verifying the issue! :+1: Your confirmation will help us to acknowledge and process this report.

**Comment by deepakk-webkul** [comment]
@sanjaychouhan-webkul why i am not able to add confirmed label on it

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @sanjaychouhan-webkul
Thank you for verifying the issue! :+1: Your confirmation will help us to acknowledge and process this report.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @AlexWorking 
 Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-99714`, `MAGETWO-99715` were created

**Issue Available**: @AlexWorking, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by ghost** [comment]
There is a similar issue https://github.com/magento/magento2/issues/22762 which I have fixed.

**Comment by sdzhepa** [comment]
The related internal Jira ticket `MC-24359` was closed as `non-reproducible` in `2.4-develop`.
It means that Magento team either unable to reproduce this issue using provided _Steps to Reproduce_ from the **_Description section_** on clean Magento instance or the issue has been already fixed in the scope of other tasks.

But if you still run into this problem on the latest `2.4-develop` please create new Issue with updated information/steps/preconditions in the **_Description section_**.

**Comment by wojtekn** [comment]
@sdzhepa did you check it on Magento Commerce?
